accord-core/src/main/java/accord/utils/ArrayBuffers.java

```diff
@@ -0,0 +1,548 @@
+        {
+            return buffer;
+        }
+
+        public int lengthOfLast(T[] buffer)
+        {
+            return buffer.length;
+        }
+
+        public boolean discard(T[] buffer, int size)
+        {
+            return discardInternal(buffer, buffer.length, size, false);
+        }
+
+        @Override
+        public boolean forceDiscard(T[] buffer, int size)
+        {
+            return discardInternal(buffer, buffer.length, size, true);
+        }
+
+        @Override
+        public T[] get(int minSize)
+        {
+            return getInternal(minSize);
+        }
+    }
+
+    /**
+     * Returns the buffer to the caller, saving the length if necessary
+     */
+    public static class PassThroughObjectBuffers<T> implements ObjectBuffers<T>
+    {
+        final ObjectBuffers<T> objs;
+        T[] savedObjs; // permit saving of precisely one unused buffer of any size to assist LinearMerge
+        int length;
+
+        public PassThroughObjectBuffers(ObjectBuffers<T> objs)
+        {
+            this.objs = objs;
+        }
+
+        @Override
+        public T[] get(int minSize)
+        {
+            length = -1;
+            if (savedObjs != null && savedObjs.length >= minSize)
+            {
+                T[] result = savedObjs;
+                savedObjs = null;
+                return result;
+            }
+            return objs.get(minSize);
+        }
+
+        @Override
+        public T[] complete(T[] buffer, int size)
+        {
+            length = size;
+            return buffer;
+        }
+
+        @Override
+        public T[] completeWithExisting(T[] buffer, int size)
+        {
+            length = size;
+            return buffer;
+        }
+
+        /**
+         * Invoke {@link #complete(Object[], int)} on the wrapped ObjectBuffers
+         */
+        public T[] realComplete(T[] buffer, int size)
+        {
+            return objs.complete(buffer, size);
+        }
+
+        @Override
+        public boolean discard(T[] buffer, int size)
+        {
+            return true;
+        }
+
+        @Override
+        public boolean forceDiscard(T[] buffer, int size)
+        {
+            if (!objs.forceDiscard(buffer, size))
+                return false;
+
+            return discardInternal(buffer);
+        }
+
+        /**
+         * Invoke {@link #discard(Object[], int)} on the wrapped ObjectBuffers
+         */
+        public void realDiscard(T[] buffer, int size)
+        {
+            if (!objs.discard(buffer, size))
+                return;
+
+            discardInternal(buffer);
+        }
+
+        private boolean discardInternal(T[] buffer)
+        {
+            if (savedObjs != null && savedObjs.length >= buffer.length)
+                return false;
+
+            savedObjs = buffer;
+            return true;
+        }
+
+        public int lengthOfLast(T[] buffer)
+        {
+            return length;
```

Comment:
this can return `-1` which means we called without a call to `complete`, so rather than returning `-1` (which no caller handles) we should fail